tests/utilities/test_cli.py

```diff
@@ -688,10 +712,172 @@ def __init__(self, optim1: dict, optim2: dict, scheduler: dict):
+            # and the `LightningCLI` subcommand exclusion list needs to be updated
+            assert e in parameters
+
+
+def test_lightning_cli_custom_subcommand():
+    class TestTrainer(Trainer):
+        def foo(self, model: LightningModule, x: int, y: float = 1.0):
+            """
+            Sample extra function.
+
+            Args:
+                model: A model
+                x: The x
+                y: The y
+            """
+
+    class TestCLI(LightningCLI):
+        @staticmethod
+        def subcommands():
+            subcommands = LightningCLI.subcommands()
+            subcommands["foo"] = {"model"}
+            return subcommands
+
+    out = StringIO()
+    with mock.patch("sys.argv", ["any.py", "-h"]), redirect_stdout(out), pytest.raises(SystemExit):
+        TestCLI(BoringModel, trainer_class=TestTrainer)
+    out = out.getvalue()
+    assert "Sample extra function." in out
+    assert "{fit,validate,test,predict,tune,foo}" in out
+
+    out = StringIO()
+    with mock.patch("sys.argv", ["any.py", "foo", "-h"]), redirect_stdout(out), pytest.raises(SystemExit):
+        TestCLI(BoringModel, trainer_class=TestTrainer)
+    out = out.getvalue()
+    assert "A model" not in out
+    assert "Sample extra function:" in out
+    assert "--x X" in out
+    assert "The x (required, type: int)" in out
+    assert "--y Y" in out
+    assert "The y (type: float, default: 1.0)" in out
+
+
+def test_lightning_cli_run():
+    with mock.patch("sys.argv", ["any.py"]):
+        cli = LightningCLI(BoringModel, run=False)
+    assert cli.trainer.global_step == 0
     assert isinstance(cli.trainer, Trainer)
     assert isinstance(cli.model, LightningModule)
+
+    with mock.patch("sys.argv", ["any.py", "fit"]):
+        cli = LightningCLI(BoringModel, trainer_defaults={"max_steps": 1, "max_epochs": 1})
+    assert cli.trainer.global_step == 1
+    assert isinstance(cli.trainer, Trainer)
+    assert isinstance(cli.model, LightningModule)
+
+
+def test_lightning_cli_config_with_subcommand(tmpdir):
+    config = {
+        # FIXME: this is not necessary, I expected it to be
+        # "subcommand": "test",
+        "test": {"trainer": {"limit_test_batches": 1}, "verbose": True, "ckpt_path": "foobar"}
+    }
+    with mock.patch("sys.argv", ["any.py", f"--config={config}"]), mock.patch(
+        "pytorch_lightning.Trainer.test", autospec=True
+    ) as test_mock:
+        cli = LightningCLI(BoringModel)
+
+    test_mock.assert_called_once_with(cli.trainer, cli.model, verbose=True, ckpt_path="foobar")
+    assert cli.trainer.limit_test_batches == 1
+
+
+def test_lightning_cli_config_before_subcommand():
+    # FIXME: passing a config with the subconfigs of multiple subcommands does not work
+    config = {
+        "test": {"trainer": {"limit_test_batches": 1}, "verbose": True, "ckpt_path": "foobar"},
+        "validate": {"trainer": {"limit_val_batches": 1}, "verbose": False, "ckpt_path": "barfoo"},
+    }
```

Comment:
@mauvilsa A very nice feature would be supporting what this test is trying to test.

That is, quickly swapping subcommands without swapping the config file, for example:

`python scrip.py --config=a_config_with_all_defaults_per_subcommand.yaml validate`
`python scrip.py --config=a_config_with_all_defaults_per_subcommand.yaml test`

Instead of the usual:

`python script.py validate --config=validate_config.yaml`
`python script.py test --config=test_config.yaml`

It currently fails with the following message:

```python
any.py: error: 'Configuration check failed :: Only values from a single sub-command are allowed ("test", "validate").'
```

Reply:
When subcommand settings are given in a config file, explicitly specifying which subcommand is not mandatory in the config or in the command line. It would be valid to do `cli.py --config=config.yaml` and no more. The version now in master if the subcommand is explicitly specified, then it just ignores the other subcommand settings instead of failing.

Reply:
> if the subcommand is explicitly specified, then it just ignores the other subcommand settings instead of failing.

Awesome!

What is missing is being able to switch the subcommand via command line, without needing to edit `subcommand: validate` in the config file.

See the test for an example, still fails with

`'Key "subcommand" is required if multiple sub-command values are provided: [\'validate\', \'test\'].'`

It only works if I change the config before running the CLI:

`config["subcommand"] = "validate"`